runtime/doc/lua.txt

```diff
@@ -2773,4 +2808,367 @@ range({spec})                                            *vim.version.range()*
+    return value is used as the table key and the second return value as the
+    table value.
+
+    Example: >lua
+
+     local it1 = vim.iter(string.gmatch('100 20 50', 'd+')):filter_map(tonumber)
+     it1:collect()
+     -- { 100, 20, 50 }
+
+     local it2 = vim.iter(string.gmatch('100 20 50', 'd+')):filter_map(tonumber)
+     it2:collect({ sort = true })
+     -- { 20, 50, 100 }
+<
+
+    Parameters: ~
+      • {opts}  ?table Optional arguments:
+                • sort (boolean|function): If true, sort the resulting table
+                  before returning. If a function is provided, that function
+                  is used as the comparator function to |table.sort()|.
+
+    Return: ~
+        (table)
+
+Iter:enumerate({self})                                      *Iter:enumerate()*
+    Add an iterator stage that returns the current iterator count as well as
+    the iterator value.
+
+    Example: >lua
+
+     local it = vim.iter(vim.gsplit('abc', '')):enumerate()
+     it:next()
+     -- 1   'a'
+     it:next()
+     -- 2   'b'
+     it:next()
+     -- 3   'c'
+<
+
+    Return: ~
+        Iter
+
+Iter:filter({self}, {f})                                       *Iter:filter()*
+    Add a filter step to the iterator pipeline.
+
+    Example: >lua
+
+     local bufs = vim.iter(vim.api.nvim_list_bufs()):filter(vim.api.nvim_buf_is_loaded)
+<
+
+    Parameters: ~
+      • {f}  function(...):bool Takes all values returned from the previous
+             stage in the pipeline and returns false or nil if the current
+             iterator element should be removed.
+
+    Return: ~
+        Iter
+
+Iter:filter_map({self}, {f})                               *Iter:filter_map()*
+    Add a filter/map step to the iterator pipeline.
+
+    Example: >lua
+
+     local it = vim.iter({ 1, 2, 3, 4 }):filter_map(function(v)
+       if v % 2 == 0 then
+         return v * 3
+       end
+     end)
+     it:collect()
+     -- { 6, 12 }
+<
+
+    Parameters: ~
+      • {f}  function(...):any Mapping function. Takes all values returned
+             from the previous stage in the pipeline as arguments and returns
+             one or more new values, which are used in the next pipeline
+             stage. Nil return values returned are filtered from the output.
+
+    Return: ~
+        Iter
+
+Iter:find({self}, {f})                                           *Iter:find()*
+    Find the first value in the iterator that satisfies the given predicate.
+
+    Advances the iterator. Returns nil and drains the iterator if no value is
+    found.
+
+    Examples: >lua
+
+     local it = vim.iter({ 3, 6, 9, 12 })
+     it:find(12)
+     -- 12
+
+     local it = vim.iter({ 3, 6, 9, 12 })
+     it:find(20)
+     -- nil
+
+     local it = vim.iter({ 3, 6, 9, 12 })
+     it:find(function(v) return v % 4 == 0 end)
+     -- 12
+<
+
+    Return: ~
+        any
+
+Iter:foreach({self}, {f})                                     *Iter:foreach()*
```

Comment:
Should it maybe be `for_each`? It seems like a lot of different languages use either camelCase `forEach` or snake_case `for_each`. (Rust, JavaScript, Java, C++, etc...). There are probably different languages that uses `foreach` as well, but IMO it looks better as `for_each` as it makes the most sense as two separate words (for each) vs (foreach)

Reply:
On the other hand, Lua tends to use single words instead of underscores:

* `tonumber`
* `tostring`
* `collectgarbage`
* `setmetatable`
* `loadstring`

etc.

Reply:
Ah yes, that's a good argument. It also seems that there are some languages that use `foreach`. It just feels a bit wrong to me, but I don't feel _super_ strongly about it, so fine with me to leave it as well

Reply:
We should be consistent either way though, so either all functions should use `singleword` style or all should use `snake_case` style. I changed everything to use `singleword` style in 39f5a075bf618af730bf6667d7d5bf85cd164951 as that is "more Lua", but if the Bikeshed Brigade comes to a consensus on `snake_case` I'll change them all back (I don't feel strongly about it either way, as long as we're consistent).

Reply:
My vote is for noodle case, for the following reason: we use snake case specifically for `verb_noun` idioms, and none of these functions have this "complex syntax" -- they're simple terms (like `foreach`, which is indeed a builtin keyword in some languages).

Reply:
I like `each` too 👍